In [1]:
import numpy as np
import pandas as pd

In [2]:
adani=pd.read_csv("ADANIPORTS.csv")
adani['return']=adani['Close'].pct_change()
adani=adani['return'].dropna()

In [3]:
cipla=pd.read_csv("CIPLA.csv")
cipla['return']=cipla['Close'].pct_change()
cipla=cipla['return'].dropna()

In [4]:
ongc=pd.read_csv("ONGC.csv")
ongc['return']=ongc['Close'].pct_change()
ongc=ongc['return'].dropna()

In [5]:
def func(data):
    n = len(data)
    mean = sum(data) / n
    deviations = sum([(x - mean)**2 for x in data])
    variance = deviations / (n - 1)
    s = variance**(1/2)
    return mean,s

In [6]:
expectedA,stddevA=func(adani)
expectedB,stddevB=func(cipla)
expectedC,stddevC=func(ongc)


In [7]:
print("AdaniPort Expected Return: "+str(expectedA))
print("Cipla Expected Return: "+str(expectedB))
print("ONGC Expected Return: "+str(expectedC))
print("AdaniPort Standard Deviation: "+str(stddevA))
print("Cipla Standard Deviation: "+str(stddevB))
print("ONGC Standard Deviation: "+str(stddevC))

AdaniPort Expected Return: 0.0005256802216729013
Cipla Expected Return: 0.0003245075403814095
ONGC Expected Return: 0.00029434814653220217
AdaniPort Standard Deviation: 0.03036997137637826
Cipla Standard Deviation: 0.0243825665272633
ONGC Standard Deviation: 0.02625614630154568


Lets suppose we invest x,y,z percent in adaniport,cipla,ongc respectively.
Risk Free return is taken to be Zero.

In [8]:
def func(r,sign=-1.0):
    x=r[0]
    y=r[1]
    z=r[2]
    Sharpe_ratio=(x*expectedA+y*expectedB+z*expectedC)/np.sqrt((x**2)*(stddevA**2)+(y**2)*(stddevB**2)+(z**2)*(stddevC**2))
    return sign*Sharpe_ratio

In [9]:
def constraint(x):
    return x[0]+x[1]+x[2]-1

In [10]:
r0=[1/3,1/3,1/3]
bnds=((0.3,1),(0.3,1),(0.3,1))
cons=[{'type':'eq','fun':constraint}]

In [11]:
from scipy.optimize import minimize

In [12]:
sol=minimize(func,r0,method='SLSQP',bounds=bnds,constraints=cons)
print(sol)

     fun: -0.024516620030793728
     jac: array([-0.00084868, -0.00058005,  0.00167406])
 message: 'Optimization terminated successfully'
    nfev: 16
     nit: 4
    njev: 4
  status: 0
 success: True
       x: array([0.35805258, 0.34194742, 0.3       ])


So best Sharpe Ratio will be 0.024516620030793728.

35.805258% in AdaniPort

34.194742% in Cipla

30% in ONGC